crates/cpuid-utils/src/lib.rs

```diff
@@ -23,55 +63,475 @@ pub mod bits;
+    ///   replaced by the supplied `values`.
+    /// - `Err` if the insert would cause the selected leaf to have one entry
+    ///   with no subleaf and one entry with a subleaf.
+    pub fn insert(
+        &mut self,
+        ident: CpuidIdent,
+        values: CpuidValues,
+    ) -> CpuidMapInsertResult {
+        match ident.subleaf {
+            Some(sl) => self.insert_leaf_subleaf(ident.leaf, sl, values),
+            None => self.insert_leaf_no_subleaf(ident.leaf, values),
+        }
+    }
+
+    /// Removes the entry with the supplied `ident` from the map if it is
+    /// present, returning its value.
+    ///
+    /// If `ident.subleaf` is `None`, this routine will only match leaf entries
+    /// that don't specify per-subleaf values.
+    pub fn remove(&mut self, ident: CpuidIdent) -> Option<CpuidValues> {
+        // If the leaf isn't present there's nothing to return.
+        let Entry::Occupied(mut entry) = self.0.entry(ident.leaf) else {
+            return None;
+        };
+
+        let (val, remove_leaf) = {
+            match (ident.subleaf, entry.get_mut()) {
+                // The caller didn't supply a subleaf, and this leaf doesn't
+                // have a subleaf map. Yank the leaf entry and return the
+                // associated value. (This can't be done inline because the
+                // entry is mutably borrowed.)
+                (None, Subleaves::Absent(val)) => (*val, true),
+                // The caller didn't supply a subleaf, but the leaf has one, so
+                // the keys don't match.
+                (None, Subleaves::Present(_)) => {
+                    return None;
+                }
+                // The caller supplied a subleaf, but the leaf doesn't use
+                // subleaves, so the keys don't match.
+                (Some(_), Subleaves::Absent(_)) => {
+                    return None;
+                }
+                // The caller supplied a subleaf, and this leaf has subleaf
+                // data. If the requested subleaf is in the leaf's subleaf map,
+                // remove the corresponding subleaf entry. If this empties the
+                // subleaf map, also clean up the leaf entry.
+                (Some(subleaf), Subleaves::Present(sl_map)) => {
+                    let val = sl_map.remove(&subleaf)?;
+                    (val, sl_map.is_empty())
+                }
+            }
+        };
+
+        if remove_leaf {
+            entry.remove_entry();
+        }
+
+        Some(val)
+    }
+
+    /// Removes all data for the supplied `leaf`. If the leaf has subleaves, all
+    /// their entries are removed.
+    pub fn remove_leaf(&mut self, leaf: u32) {
+        self.0.remove(&leaf);
+    }
+
+    /// Passes each leaf number in the map to `f` and removes any leaf entries
+    /// for which `f` returns `false`. If a removed leaf has subleaves, all
+    /// their entries are removed.
+    pub fn retain_by_leaf<F>(&mut self, mut f: F)
```

Comment:
It seems like it could, perhaps, be useful to also pass the value into `f` so that it could retain based on which subleaves are present --- but that would require making the `Subleaves` enum public, and perhaps we have no need for such an operation currently...